### PR TITLE
Added Augmentation Operations and unit tests

### PR DESCRIPTION
**Pull request by kardevroop**
Please review the changes.

### CODE REVIEW

File: ezflow/data/dataset/base_dataset.py [hunk 1]
```diff
@@ -133,6 +137,15 @@ def __getitem__(self, index):
         img2 = torch.from_numpy(img2).permute(2, 0, 1).float()
         flow = torch.from_numpy(flow).permute(2, 0, 1).float()
 
+        if self.normalize:
+            #input_transform = transforms.Compose([
```
Comment (prajnan93) [inline ezflow/data/dataset/base_dataset.py:141]: Can we remove these comments if not required for future use?

File: ezflow/data/dataset/base_dataset.py [hunk 2]
```diff
@@ -133,6 +137,15 @@ def __getitem__(self, index):
...
+            #transforms.Normalize(mean=[0,0,0], std=[255,255,255]),
+            #])
+            #img1 = input_transform(img1)
+            #img2 = input_transform(img2)
+            img1 = normalize_image(img1)
```
Comment (prajnan93) [inline ezflow/data/dataset/base_dataset.py:146]: let's make this operation in a single step.

For example: `img1, img2 = normalize_image(img1, img2)`

File: ezflow/functional/data_augmentation/operations.py [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import cv2
 import numpy as np
 from PIL import Image
+import scipy.ndimage as ndimage
 from torchvision.transforms import ColorJitter
-
+import torchvision.transforms as transforms
```
Comment (prajnan93) [inline ezflow/functional/data_augmentation/operations.py:6]: The import statements look out of order. Please refer to the CONTRIBUTING.rst file on how to set up pre-commit hooks. This will auto format the code.

File: ezflow/functional/data_augmentation/augmentor.py [hunk 4]
```diff
@@ -64,6 +66,12 @@ def __call__(self, img1, img2, flow, valid=None):
         img1, img2, flow = spatial_transform(
             img1, img2, flow, self.crop_size, **self.spatial_aug_params
         )
+        img1, img2, flow = translate_transform(
+            img1, img2, flow, **self.affine_params
```
Comment (prajnan93) [inline ezflow/functional/data_augmentation/augmentor.py:70]: Should we keep two separate affine_params? For example: `translate_affine_params` and `rotation_affine_params` .  Users might want to configure different values for `translate`, `degree` and `delta`. Let me know what your thoughts.

File: ezflow/functional/data_augmentation/operations.py [hunk 5]
```diff
@@ -400,3 +401,158 @@ def sparse_spatial_transform(
...
+        img2 = ndimage.interpolation.rotate(img2, angle2, reshape=False, order=2)
+        flow = ndimage.interpolation.rotate(flow, angle1, reshape=False, order=2)
+        
+        target_ = np.copy(flow)
+        flow[:,:,0] = np.cos(angle1_rad)*target_[:,:,0] + np.sin(angle1_rad)*target_[:,:,1]
+        flow[:,:,1] = -np.sin(angle1_rad)*target_[:,:,0] + np.cos(angle1_rad)*target_[:,:,1]
+
+    return img1, img2, flow
+
+def normalize_image(img):
```
Comment (prajnan93) [inline ezflow/functional/data_augmentation/operations.py:542]: It would be better to provide user configurable `mean` and `std` and their corresponding default values.

File: ezflow/data/dataset/base_dataset.py [hunk 6]
```diff
@@ -46,8 +47,10 @@ def __init__(
             "color_aug_params": {"aug_prob": 0.2},
             "eraser_aug_params": {"aug_prob": 0.5},
             "spatial_aug_params": {"aug_prob": 0.8},
+            "affine_params": {"aug_prob": 0.8},
         },
         sparse_transform=False,
+        normalize=False,
```
Comment (prajnan93) [inline ezflow/data/dataset/base_dataset.py:53]: It will be better to provide this feature as user configurable. 

For example: normalize={"use": True, mean=[0,0,0], std=[255.0,255.0,255.0]}

File: ezflow/functional/data_augmentation/augmentor.py (hunk 4 above)
Comment (kardevroop) [reply ezflow/functional/data_augmentation/augmentor.py:70]: I'll see if it can be done. Will add a bit more complexity in regards to the yaml file and default argument values.

File: ezflow/data/dataset/base_dataset.py (hunk 1 above)
Comment (kardevroop) [reply ezflow/data/dataset/base_dataset.py]: Comments have been removed.

File: ezflow/data/dataset/base_dataset.py (hunk 2 above)
Comment (kardevroop) [reply ezflow/data/dataset/base_dataset.py]: Done.

File: ezflow/functional/data_augmentation/augmentor.py (hunk 4 above)
Comment (kardevroop) [reply ezflow/functional/data_augmentation/augmentor.py]: Done

File: ezflow/functional/data_augmentation/operations.py (hunk 5 above)
Comment (kardevroop) [reply ezflow/functional/data_augmentation/operations.py]: Done as suggested.

File: ezflow/data/dataset/base_dataset.py (hunk 6 above)
Comment (kardevroop) [reply ezflow/data/dataset/base_dataset.py]: Done

**Comment by kardevroop** [comment]
Changes have been done. I will run a model just once to check if everything checks out. Will update.

File: ezflow/functional/data_augmentation/operations.py (hunk 3 above)
Comment (kardevroop) [reply ezflow/functional/data_augmentation/operations.py:4]: Done.

**Comment by kardevroop** [comment]
@prajnan93 the changes have been implemented. Please have a look.

File: ezflow/functional/data_augmentation/operations.py [hunk 7]
```diff
@@ -400,3 +402,173 @@ def sparse_spatial_transform(
...
+    -----------
+    USE: boolean
+        Whether to normalize image or not
+    MEAN: list
+        The list of mean values to be substracted from each image channel
+    STD: list
+        The list of std values with which to divide each image channel by
+    """
+
+    def __init__(self, USE=False, MEAN=[0, 0, 0], STD=[255.0, 255.0, 255.0]):
```
Comment (prajnan93) [inline ezflow/functional/data_augmentation/operations.py:559]: let's use lower cases for argument names to keep it consistent for the rest of the library.

File: ezflow/data/dataset/base_dataset.py [hunk 8]
```diff
@@ -101,6 +106,17 @@ def __getitem__(self, index):
         img1 = np.array(img1).astype(np.uint8)
         img2 = np.array(img2).astype(np.uint8)
 
+        img1 = torch.from_numpy(img1).permute(2, 0, 1).float()
```
Comment (prajnan93) [inline ezflow/data/dataset/base_dataset.py:109]: let's reduce the duplicate operations of converting to and from numpy to pytorch. We can move the code to make grayscale image to rgb at `line:120` before `line 109`

File: ezflow/data/dataset/base_dataset.py [hunk 9]
```diff
@@ -101,6 +106,17 @@ def __getitem__(self, index):
...
+        img2 = torch.from_numpy(img2).permute(2, 0, 1).float()
+
+        img1, img2 = self.normalize(img1, img2)
```
Comment (prajnan93) [inline ezflow/data/dataset/base_dataset.py:112]: We can move the `normalize` operation to `line no: 152` and also add it` line no:131`

Comment (kardevroop) [reply ezflow/data/dataset/base_dataset.py]: Done

File: ezflow/data/dataset/base_dataset.py (hunk 8 above)
Comment (kardevroop) [reply ezflow/data/dataset/base_dataset.py]: Done

File: ezflow/functional/data_augmentation/operations.py (hunk 7 above)
Comment (kardevroop) [reply ezflow/functional/data_augmentation/operations.py]: Done

**Comment by kardevroop** [comment]
@prajnan93  Changes have been implemented. Currently running a few epochs to make sure everything is ok.

**Review by prajnan93 — APPROVED** [review]
Looks good!
